### PR TITLE
Refactor mgmt and filter

### LINKED ISSUE #56: Provide query filters for /apis and /tenants endpoints
Would like to have ability to filter the returned data.  Syntax in the following proposed examples is not final, but conveys the kind of filtering being requested.

`GET /tenants?namespace=NAMESPACE`
`GET /tenants?namespace=NAMESPACE&instance=INSTANCE`
`GET /tenants/TENANTID/apis?basePath=BASEPATH`
`GET /tenants/TENANTID/apis?name=NAME`

### PR DESCRIPTION
**Pull request by alexsong93**
Separated out the management interface to different files for APIs, Tenants, Resources, and Subscriptions. Also implemented filtering for APIs and Tenants. Closes #56. @lostinthestory @mhamann PTAL

### CODE REVIEW

File: api-gateway-config/conf.d/management_apis.conf [hunk 1]
```diff
@@ -35,16 +35,16 @@ server {
 
     location /v1/apis {
         access_by_lua_block {
-            local mgmt = require("management")
+            local apis = require("management/apis")
```
Comment (ddragosd) [inline api-gateway-config/conf.d/management_apis.conf]: WDYT if we move these lines inside the `management/apis` module and keep the nginx conf at the minimum with the amount of lua code. I.e.

```nginx
location /v1/apis {
  content_by_lua_block {
    local apis = require("management/apis")
    apis.requestHandler()
  }
}
```

Comment (ddragosd) [inline api-gateway-config/conf.d/management_apis.conf]: ... and maybe do the same for `/v1/tenants` and `/v1/subscriptions` ?

File: api-gateway-config/scripts/lua/management/resources.lua [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+-- @param red
+-- @param resource
+-- @param gatewayPath
+-- @param tenantId
+function _M.addResource(red, resource, gatewayPath, tenantId)
+  -- Create resource object and add to redis
+  local redisKey = utils.concatStrings({"resources", ":", tenantId, ":", gatewayPath})
+  local apiId
+  local operations
+  for k, v in pairs(resource) do
```
Comment (ddragosd) [inline api-gateway-config/scripts/lua/management/resources.lua]: do we need the loop here ?
IIUC `resource` comes as a Lua table/object and we could use:
`resource.apiId` or `resources.operations` ?

File: api-gateway-config/scripts/lua/management/subscriptions.lua [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  -- Read in the PUT JSON Body
+  ngx.req.read_body()
+  local args = ngx.req.get_body_data()
+  if not args then
+    request.err(400, "Missing request body.")
+  end
+  -- Convert json into Lua table
+  local decoded = cjson.decode(args)
+  -- Check required fields
+  local requiredFieldList = {"key", "scope", "tenantId"}
```
Comment (ddragosd) [inline api-gateway-config/scripts/lua/management/subscriptions.lua]: I think we can extract the input validation to make it reusable. I'm seeing the same logic in other parts, i.e. https://github.com/openwhisk/apigateway/pull/62/files#diff-eca0bb209f1cd0d1f6c13768183f866eR62

File: api-gateway-config/scripts/lua/management/tenants.lua [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+end
+
+--- Check for tenant id from uri and use existing tenant if it already exists in redis
+-- @param red Redis client instance
+-- @param uri Uri of request. Eg. /v1/tenants/{id}
+function checkURIForExistingTenant(red, uri)
+  local id, existing
+  local index = 1
+  -- Check if id is in the uri
+  for word in string.gmatch(uri, '([^/]+)') do
```
Comment (ddragosd) [inline api-gateway-config/scripts/lua/management/tenants.lua]: I think we can reduce this code to:
```lua
ngx.var.tenant_id
```
provided we configure an NGINX location with named regular expression. I.e.
```nginx
location ~ /v1/tenants/?(?<tenant_id>.*)? {
    # variable $tenant_id is now available
}
```

File: api-gateway-config/scripts/lua/management/tenants.lua [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+--- Get one or all tenants from the gateway
+-- GET /v1/tenants
+function _M.getTenants()
+  local uri = string.gsub(ngx.var.request_uri, "?.*", "")
+  local queryParams = ngx.req.get_uri_args()
+  local id
+  local index = 1
+  local apiQuery = false
+  for word in string.gmatch(uri, '([^/]+)') do
```
Comment (ddragosd) [inline api-gateway-config/scripts/lua/management/tenants.lua]: see the comment on : https://github.com/openwhisk/apigateway/pull/62/files#diff-eca0bb209f1cd0d1f6c13768183f866eR89

File: api-gateway-config/scripts/lua/management/tenants.lua [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+  end
+  ngx.header.content_type = "application/json; charset=utf-8"
+  request.success(200, tenantList)
+end
+
+--- Filter tenants based on query parameters
+-- @param tenants list of tenants
+-- @param queryParams query parameters to filter tenants
+function filterTenants(tenants, queryParams)
+  local namespace = queryParams['filter[where][namespace]']
```
Comment (ddragosd) [inline api-gateway-config/scripts/lua/management/tenants.lua]: IIRC another way to read query params could be `ngx.var.arg_<param_name>`

Comment (alexsong93) [inline api-gateway-config/scripts/lua/management/tenants.lua]: Yes. I implemented it this way because `queryParams` is used elsewhere in the code as well.

**Comment by alexsong93** [comment]
@ddragosd Thanks for your input! I added a new commit based on the comments.

**Comment by ddragosd** [comment]
thanks @alexsong93 . Are the existing test cases covering for these changes too ?

**Comment by alexsong93** [comment]
@ddragosd most of the changes here are for the management interface, which is covered by the OpenWhisk test pipeline. We're also looking into adding more comprehensive integration tests, and I believe @taylorking started working on it in #67.

**Pull request merged**